### PR TITLE
QL: Refactor FunctionRegistry to make it pluggable

### PR DESCRIPTION
**Pull request by costin**
Break the FunctionRegistry monolith into a common QL base and a SQL
specific registry that handles aspects such as distinct and extract.
In the process clean-up the names and signature of internal interfaces.
Most of the semantics were preserved however the error messages were
slightly tweaked to make them more readable - this shouldn't be a
problem as they are being used internally mainly in test assertions.

### CODE REVIEW

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionDefinition.java [hunk 1]
```diff
@@ -17,28 +17,19 @@
      */
     @FunctionalInterface
     public interface Builder {
-        Function build(UnresolvedFunction uf, boolean distinct, Configuration configuration);
+        Function build(UnresolvedFunction uf, Configuration configuration, Object... extras);
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionDefinition.java:20]: This is the entry point of this PR - the varargs allow subclasses to specify extra parameters however it's up to them to do routing and verify parameters.

Comment (matriv) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionDefinition.java:20]: Could you maybe add some more details to the javadoc of the Builder?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java [hunk 2]
```diff
@@ -274,11 +296,147 @@ public SqlFunctionRegistry() {
...
+        };
+        return new SqlFunctionDefinition(primaryName, unmodifiableList(aliases), function, datetime, realBuilder);
+    }
+
+    private static Boolean asBool(Object[] extras) {
+        if (CollectionUtils.isEmpty(extras)) {
+            return null;
+        }
+        if (extras.length != 1 || (extras[0] instanceof Boolean) == false) {
+            throw new SqlIllegalArgumentException("Invalid number and types of arguments given to function definition");
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java:340]: Does it make sense to be more explicit in the error message here? The message says it's an invalid number and type of arguments but it doesn't say what is expecting.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java [hunk 3]
```diff
@@ -274,11 +296,147 @@ public SqlFunctionRegistry() {
...
+            if (children.size() != 3) {
+                throw new QlIllegalArgumentException("expects three arguments");
+            }
+            forbidDistinct(source, distinct);
+            return ctorRef.build(source, children.get(0), children.get(1), children.get(2), cfg.zoneId());
+        };
+        return def(function, builder, true, names);
+    }
+
+    protected interface TertiaryZoneIdAwareBuilder<T> {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java:418]: For a three-argument function here you use `tertiary` naming, while in `FunctionRegistry` you use `ternary`. For an uniform approach, could you use the same style in both places? If so, given the rest of the names - unary, binary -, I'd say it should be `ternary`.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java [hunk 4]
```diff
@@ -274,11 +296,147 @@ public SqlFunctionRegistry() {
...
+        T build(Source source, Expression first, Expression second, Expression third, ZoneId zi);
+    }
+
+
+    /**
+     * Special method to create function definition for Cast as its signature is not compatible with {@link UnresolvedFunction}.
+     */
+    @SuppressWarnings("overloads")  // These are ambiguous if you aren't using ctor references but we always do
+    protected static <T extends Function> FunctionDefinition def(Class<T> function, CastBuilder<T> ctorRef, String... names) {
+        SqlFunctionBuilder builder = (source, children, cfg, distinct) ->
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java:428]: Shouldn't `forbidDistinct` be used here as well?

**Review by astefan — APPROVED** [review]
LGTM
Also, left few minor comments.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java [hunk 5]
```diff
@@ -124,304 +125,133 @@ public boolean functionExists(String functionName) {
...
     @SuppressWarnings("overloads")  // These are ambiguous if you aren't using ctor references but we always do
-    public static <T extends Function> FunctionDefinition def(Class<T> function,
-            BinaryFunctionBuilder<T> ctorRef, String... names) {
-        FunctionBuilder builder = (source, children, distinct, cfg) -> {
+    protected static <T extends Function> FunctionDefinition def(Class<T> function, BinaryBuilder<T> ctorRef, String... names) {
+        FunctionBuilder builder = (source, children, cfg) -> {
             boolean isBinaryOptionalParamFunction = OptionalArgument.class.isAssignableFrom(function);
             if (isBinaryOptionalParamFunction && (children.size() > 2 || children.size() < 1)) {
                 throw new QlIllegalArgumentException("expects one or two arguments");
             } else if (!isBinaryOptionalParamFunction && children.size() != 2) {
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java:215]: I didn't see these statements and frankly I was more concerned about touching extra code that leads to extra formatting and extra noise.

As for the consistent approach, we could reinforce it project wide which could align with the spotless/formatting discussion which I believe is scheduled for Elasticsearch 8.

[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java]: Btw, see https://github.com/elastic/elasticsearch/pull/67817

**Comment by costin** [comment]
@elasticmachine update branch

**Review by palesz — APPROVED** [review]
LGTM, two minor comments

**Pull request merged**